In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import re
from collections import Counter
import plotly.graph_objects as go

In [2]:
df = pd.read_csv(r"C:\Users\Suwani Gunasekara\OneDrive\Documents\Sem 7 ML\Mid term\results.csv")

In [3]:
df.head()

date home_team away_team  home_score  away_score tournament     city  \
0  1872-11-30  Scotland   England           0           0   Friendly  Glasgow   
1  1873-03-08   England  Scotland           4           2   Friendly   London   
2  1874-03-07  Scotland   England           2           1   Friendly  Glasgow   
3  1875-03-06   England  Scotland           2           2   Friendly   London   
4  1876-03-04  Scotland   England           3           0   Friendly  Glasgow   

    country  neutral  
0  Scotland    False  
1   England    False  
2  Scotland    False  
3   England    False  
4  Scotland    False

In [4]:
# getting the team that won
def teamWon(df):
    if(df['home_score']>df['away_score']):
        return df.home_team
    elif(df['home_score']<df['away_score']):
        return df.away_team
    elif(df['home_score']==df['away_score']):
        return 0  

In [5]:
df["TeamWon"] = df.apply(teamWon,axis = 1)

In [6]:
# Converting df column to list
listwon = df.TeamWon.values.tolist()

In [36]:
# Getting the number of wins of each team and converting it to a dictionary 
# https://stackoverflow.com/questions/11068986/how-to-convert-counter-object-to-dict
dictWon = dict(Counter(listwon))

In [37]:
# creating two new lists to put countries who got draws
homedraws = []
awaydraws = []

In [38]:
# appending to the list
for i in range(len(df.TeamWon)):
    if df.TeamWon[i] == 0:
        homedraws.append(df.home_team[i])
        awaydraws.append(df.away_team[i])

In [39]:
# Creating dictionaries 
homedrawsdict = dict(Counter(homedraws))
awaydrawsdict = dict(Counter(awaydraws))

In [40]:
#https://stackoverflow.com/questions/10461531/merge-and-sum-of-two-dictionaries
alldrawsdict = dict(Counter(homedrawsdict)+ Counter(awaydrawsdict))

In [41]:
#alldrawsdict

In [42]:
# Number of home games played
homegamesdict = dict(Counter(df.home_team.values.tolist()))
# Number of away games played
awaygamesdict = dict(Counter(df.away_team.values.tolist()))
# Number of total games played
totalgamesdict = dict(Counter(homegamesdict)+Counter(awaygamesdict))

In [43]:
# Updating alldrawsdict to contain zero for countries which doesnt have draws
for k,v in totalgamesdict.items():
    if k not in alldrawsdict:
        alldrawsdict[k] = 0

In [44]:
# Updating dictWon to contain zero for countries which doesnt have wins
for k,v in totalgamesdict.items():
    if k not in dictWon:
        dictWon[k] = 0

In [45]:
#wondict = {}
#for k,v in totalgamesdict.items():
#    try:
#        wondict[k] = (dictWon[k]+0.5*alldrawsdict[k])/totalgamesdict[k]
#    except KeyError:
#        print(k)

In [46]:
#for k in sorted(wondict,key =wondict.get, reverse = True):
#    print(k,"--->",wondict[k])


In [47]:
# get the total number of games played
totalmatches = 0
for k,v in totalgamesdict.items():
    totalmatches += totalgamesdict[k]

In [48]:
# average games played by a team
averageGames = totalmatches/len(totalgamesdict)
print(averageGames)

269.7564102564103


In [49]:
wonratedict = {}
for k,v in totalgamesdict.items():
    if v >= averageGames:
        try:
            wonratedict[k] = (dictWon[k]+0.5*alldrawsdict[k])/totalgamesdict[k]
        except KeyError:
            print(k)

In [50]:
for k in sorted(wonratedict,key =wonratedict.get, reverse = True):
    print(k,"--->",wonratedict[k])

Brazil ---> 0.7395939086294416
Spain ---> 0.6985815602836879
England ---> 0.6877450980392157
Germany ---> 0.6857440166493236
Iran ---> 0.6752988047808764
Italy ---> 0.6705665024630542
Argentina ---> 0.6615853658536586
Croatia ---> 0.6595744680851063
South Korea ---> 0.6543352601156069
Russia ---> 0.6418439716312057
Czech Republic ---> 0.6363636363636364
Ivory Coast ---> 0.6239015817223199
Mexico ---> 0.6211428571428571
Netherlands ---> 0.6206030150753769
Egypt ---> 0.6132075471698113
Morocco ---> 0.6103896103896104
Ghana ---> 0.609271523178808
Nigeria ---> 0.6077586206896551
Australia ---> 0.6054158607350096
France ---> 0.6051703877790834
Iraq ---> 0.605072463768116
Portugal ---> 0.6033653846153846
Sweden ---> 0.6004854368932039
Japan ---> 0.5963665086887836
China PR ---> 0.5947986577181208
Saudi Arabia ---> 0.5923566878980892
Ukraine ---> 0.5899280575539568
Zambia ---> 0.5899031811894883
Cameroon ---> 0.5882899628252788
Martinique ---> 0.5855263157894737
Scotland ---> 0.58164556962025

### Era dominated

In [51]:
df['BrazilWon'] = np.where(df.TeamWon == "Brazil",1,0)
df['SpainWon'] = np.where(df.TeamWon == 'Spain',1,0)
df['EnglandWon'] = np.where(df.TeamWon == 'England',1,0)
df['GermanyWon'] = np.where(df.TeamWon == 'Germany',1,0)
df['IranWon'] = np.where(df.TeamWon == 'Iran',1,0)

In [52]:
df.head()

date home_team away_team  home_score  away_score tournament     city  \
0  1872-11-30  Scotland   England           0           0   Friendly  Glasgow   
1  1873-03-08   England  Scotland           4           2   Friendly   London   
2  1874-03-07  Scotland   England           2           1   Friendly  Glasgow   
3  1875-03-06   England  Scotland           2           2   Friendly   London   
4  1876-03-04  Scotland   England           3           0   Friendly  Glasgow   

    country  neutral   TeamWon  BrazilWon  SpainWon  EnglandWon  GermanyWon  \
0  Scotland    False         0          0         0           0           0   
1   England    False   England          0         0           1           0   
2  Scotland    False  Scotland          0         0           0           0   
3   England    False         0          0         0           0           0   
4  Scotland    False  Scotland          0         0           0           0   

   IranWon  
0        0  
1        0  
2        0  
3        0  
4        0

In [53]:
newdate = df["date"].str.split("-",n=1, expand = True) 
# combing year values as a new column to df
df['Year'] = newdate[0]

In [60]:
# grouping based on year and summarizing the number of matches each country won
Brazil = df.groupby(['Year'])['BrazilWon'].sum()
Spain = df.groupby(['Year'])['SpainWon'].sum()
England = df.groupby(['Year'])['EnglandWon'].sum()
Germany = df.groupby(['Year'])['GermanyWon'].sum()
Iran = df.groupby(['Year'])['IranWon'].sum()

In [64]:
dfEra  = pd.concat([Brazil,Spain,England,Germany,Iran],axis = 1).reset_index()
dfEra.head()

Year  BrazilWon  SpainWon  EnglandWon  GermanyWon  IranWon
0  1872          0         0           0           0        0
1  1873          0         0           1           0        0
2  1874          0         0           0           0        0
3  1875          0         0           0           0        0
4  1876          0         0           0           0        0

In [75]:
fig = go.Figure()
fig.add_trace(go.Scatter(x = dfEra.Year, y = dfEra.BrazilWon, name = "Brazil", line_shape = 'linear'))
fig.add_trace(go.Scatter(x = dfEra.Year, y = dfEra.SpainWon, name = "Spain", line_shape='linear'))
fig.add_trace(go.Scatter(x = dfEra.Year, y = dfEra.EnglandWon, name = "England", line_shape='linear'))
fig.add_trace(go.Scatter(x = dfEra.Year, y = dfEra.GermanyWon, name = "German", line_shape='linear'))
fig.add_trace(go.Scatter(x = dfEra.Year, y = dfEra.IranWon, name = "Iran", line_shape='linear'))

fig.update_layout(
    title="Year vs the number of goals played by each country",
    xaxis_title="Years",
    yaxis_title="Number of goals",
    legend_title="Country",
    font=dict(
        family="Courier New, monospace",
        size=12,
        color="RebeccaPurple"
    )
)

fig.show()

- Brazil dominated from 1956 to 1962
- England dominated from 1872 to 1913
